### PR TITLE
Log exceptions in tests to stderr

### PR DESCRIPTION
**Pull request by DMRobertson**
This makes it easy to see errors in background processes when you're
writing unit tests. Otherwise, you have to know/remember to look at the
test logs in `_trial_temp`.

### CODE REVIEW

File: tests/test_utils/logging_setup.py [hunk 1]
```diff
@@ -35,7 +36,8 @@ def emit(self, record):
 def setup_logging():
     """Configure the python logging appropriately for the tests.
 
-    (Logs will end up in _trial_temp.)
+    Logs will end up in _trial_temp. Exceptions are additionally
```
Comment (richvdh) [inline tests/test_utils/logging_setup.py:39]: Not everything logged at `logger.error` is an exception! `s/exceptions/errors/`.

Likewise in the changelog, and the name of the handler.

File: tests/test_utils/logging_setup.py [hunk 2]
```diff
@@ -44,11 +46,16 @@ def setup_logging():
...
     )
 
-    handler = ToTwistedHandler()
+    to_twisted_handler = ToTwistedHandler()
     formatter = logging.Formatter(log_format)
-    handler.setFormatter(formatter)
-    handler.addFilter(LoggingContextFilter())
-    root_logger.addHandler(handler)
+    to_twisted_handler.setFormatter(formatter)
+    to_twisted_handler.addFilter(LoggingContextFilter())
```
Comment (richvdh) [inline tests/test_utils/logging_setup.py:52]: I *think* this is redundant, as of #8051. (If it's not redundant, we need to do it for the other handler too.)

**Review by richvdh — CHANGES_REQUESTED** [review]

File: tests/test_utils/logging_setup.py (hunk 2 above)
Comment (DMRobertson) [reply tests/test_utils/logging_setup.py:52]: Alright. I propose to test if this is redundant by
* removing filter here
* changing LoggingContextFilter to have some obvious side effect
* trigger the new log-to-stderr behaviour, and look for the side effect

**Comment by richvdh** [comment]
I'm a bit concerned that this makes the test output hard to parse - there's actually quite a lot of stuff logged at ERROR which is expected in the tests, and this makes it look like things are failing when they're not.

Might be worth getting opinions from the rest of the team, in the room?

**Comment by richvdh** [comment]
(on the other hand, possibly a bunch of those things logged at ERROR shouldn't be...)

File: tests/test_utils/logging_setup.py (hunk 2 above)
Comment (DMRobertson) [reply tests/test_utils/logging_setup.py:52]: Appears to be needed. Otherwise the call to `debug` in synapse/metrics/__init__.py:596 will fail because the record doesn't have a "request" field.

**Comment by DMRobertson** [comment]
Don't think anyone was super keen on this, so I'll close this down.

Maybe something simpler would suffice. (If a test fails, print out "check homeserver.log in _trial_temp" in bright yellow)

**Comment by DMRobertson** [comment]
Argh. Just lost about an hour to this again.

Can we do better? I'm guessing somewhere high up we catch any remaining uncaught exceptions and `logger.exception` them. Can we stop that during test runs so that they bubble up immediately?

In the past I've used [pytest.raises](https://docs.pytest.org/en/latest/reference/reference.html#pytest.raises) to handle tests that are expected to raise exceptions.

**Comment by richvdh** [comment]
uncaught exceptions will cause a test to fail.

**Comment by richvdh** [comment]
(I think the problem is that the exceptions being logged *are* caught)

It would be instructive to see the sorts of errors that you are seeing being logged that would have been useful to make a test fail. Then we can see where it is being caught, and whether it would make sense to stop catching it during tests.
